quickjs.c

```diff
@@ -10226,29 +10233,21 @@ static JSValue js_atof2(JSContext *ctx, const char *str, const char **pp,
             radix = 2;
         } else if ((p[1] >= '0' && p[1] <= '9') &&
                    radix == 0 && (flags & ATOD_ACCEPT_LEGACY_OCTAL)) {
-            int i;
             has_legacy_octal = TRUE;
             sep = 256;
-            for (i = 1; (p[i] >= '0' && p[i] <= '7'); i++)
-                continue;
-            if (p[i] == '8' || p[i] == '9')
-                goto no_prefix;
             p += 1;
-            radix = 8;
-        } else {
-            goto no_prefix;
-        }
-        /* there must be a digit after the prefix */
-        if (to_digit((uint8_t)*p) >= radix)
+            for (i = 0; (p[i] >= '0' && p[i] <= '7'); i++)
+                continue;
+            if (p[i] != '8' && p[i] != '9')
+                radix = 8;
+        } else if (p[1] == sep) {
+            // Numeric separator can not be used after leading 0
             goto fail;
-    no_prefix: ;
-    } else {
- no_radix_prefix:
-        if (!(flags & ATOD_INT_ONLY) &&
-            atod_type == ATOD_TYPE_FLOAT64 &&
-            strstart(p, "Infinity", &p)) {
-            double d = INF;
-            if (is_neg)
+        }
+    } else {
+        if (*p == 'I' && (flags & ATOD_ACCEPT_INFINITY) && strstart(p, "Infinity", &p)) {
```

Comment:
Looking at `strstart`, wouldn't we be reading past the bounday if it just says "Inf" for example?

Reply:
`strstart` does not read beyond the end of the string pointed to by its first argument. The string passed to `js_atof` is null terminated, so no risk of UB.  The initial test `*p == 'I'` is for performance, almost always false.

Reply:
👍